From e77b876f2f2632d84f71005a6f6c8b150e691309 Mon Sep 17 00:00:00 2001
From: "Joel Fernandes (Google)" <joel@joelfernandes.org>
Date: Sun, 16 Oct 2022 16:22:58 +0000
Subject: [PATCH 06/20] FROMGIT: percpu-refcount: Use call_rcu_hurry() for
 atomic switch

call_rcu() changes to save power will slow down the percpu refcounter's
"per-CPU to atomic switch" path. The primitive uses RCU when switching to
atomic mode. The enqueued async callback wakes up waiters waiting in the
percpu_ref_switch_waitq. Due to this, per-CPU refcount users will slow down,
such as blk_pre_runtime_suspend().

Use the call_rcu_hurry() API instead which reverts to the old behavior.

(cherry picked from commit ab27dad9baf60339ff5293249a45f6ef88554536
 https://git.kernel.org/pub/scm/linux/kernel/git/paulmck/linux-rcu.git rcu/next)

Bug: 258241771
Bug: 222463781
Test: powerIdle lab tests.
Signed-off-by: Joel Fernandes (Google) <joel@joelfernandes.org>
Change-Id: If479fa3ee22258f656016e438b2ba0686196a065
Signed-off-by: Joel Fernandes <joelaf@google.com>
Reviewed-on: https://chromium-review.googlesource.com/c/chromiumos/third_party/kernel/+/3965074
Reviewed-by: Ross Zwisler <zwisler@google.com>
(cherry picked from commit b3d0bba5548556e58508241be437a640626436cb)
[Cherry picked from chromeos-5.10 tree. Minor tweaks to commit message
to match Android style]
Signed-off-by: Qais Yousef <qyousef@google.com>

Upstream: 343a72e5e37d380b70534fae3acd7e5e39adb769 ("percpu-refcount: Use call_rcu_hurry() for atomic switch")
ACK android13-5.10: 879a3f20ce1edd8db79ae786768a9da390c2ff46, taken from arter97/master
Backport notes: call_rcu_flush() renamed to call_rcu_hurry().
---
 lib/percpu-refcount.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/percpu-refcount.c b/lib/percpu-refcount.c
index 493093b97093..0a7a211c561b 100644
--- a/lib/percpu-refcount.c
+++ b/lib/percpu-refcount.c
@@ -224,7 +224,8 @@ static void __percpu_ref_switch_to_atomic(struct percpu_ref *ref,
 		percpu_ref_noop_confirm_switch;
 
 	percpu_ref_get(ref);	/* put after confirmation */
-	call_rcu(&ref->data->rcu, percpu_ref_switch_to_atomic_rcu);
+	call_rcu_hurry(&ref->data->rcu,
+		       percpu_ref_switch_to_atomic_rcu);
 }
 
 static void __percpu_ref_switch_to_percpu(struct percpu_ref *ref)
